### ISSUE TITLE
Use Nullable Reference Types helper instead of canBeNull argument in EntityMappingBuilder.Association methods

### ISSUE THREAD

**Issue by travbeamo** [linq2db/linq2db#2592]
With the introduction of Nullable Reference Types in C# 8.0, is it possible to consider refactoring the **Association** methods in  [EntityMappingBuilder](https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Mapping/EntityMappingBuilder.cs)  by removing the need for the optional **canBeNull** argument.

If the helper method is added to [MemberHelper](https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Expressions/MemberHelper.cs):
```cs
public const string NullableAttributeFullName = "System.Runtime.CompilerServices.NullableAttribute";
public const string NullableContextAttributeFullName = "System.Runtime.CompilerServices.NullableContextAttribute";

public static bool IsNullable(this MemberInfo member) {
  var type = member.GetType();
  if (type.IsValueType) 
    return Nullable.GetUnderlyingType(type) != null;
  var nullable = member.CustomAttributes.FirstOrDefault(x => x.AttributeType.FullName == NullableAttributeFullName);
  if (nullable != null && nullable.ConstructorArguments.Count == 1) {
    var attributeArgument = nullable.ConstructorArguments[0];
    if (attributeArgument.ArgumentType == typeof(byte[])) {
      var args = (ReadOnlyCollection<CustomAttributeTypedArgument>)attributeArgument.Value;
      if (args.Count > 0 && args[0].ArgumentType == typeof(byte)) {
        return (byte)args[0].Value == 2;
      }
    } else if (attributeArgument.ArgumentType == typeof(byte)) {
      return (byte)attributeArgument.Value == 2;
    }
  }
  var context = member.DeclaringType.CustomAttributes.FirstOrDefault(x => x.AttributeType.FullName == NullableContextAttributeFullName);
  if (context != null && context.ConstructorArguments.Count == 1 && context.ConstructorArguments[0].ArgumentType == typeof(byte)) {
    return (byte)context.ConstructorArguments[0].Value == 2;
  }
  // Couldn't find a suitable attribute
  return false;
}
```

If would then be possible to change the methods to below:
```cs
    public PropertyMappingBuilder<TEntity, TProperty> Association<TProperty, TThisKey, TOtherKey>(
      Expression<Func<TEntity, TProperty>> prop,
      Expression<Func<TEntity, TThisKey>> thisKey,
      Expression<Func<TProperty, TOtherKey>> otherKey
      //bool canBeNull = true
      ) {
      if (prop == null) throw new ArgumentNullException(nameof(prop));
      if (thisKey == null) throw new ArgumentNullException(nameof(thisKey));
      if (otherKey == null) throw new ArgumentNullException(nameof(otherKey));
      var thisMember = MemberHelper.GetMemberInfo(thisKey);
      var thisKeyName = thisMember.Name;
      var otherKeyName = MemberHelper.GetMemberInfo(otherKey).Name;
      return Property(prop).HasAttribute(new AssociationAttribute { ThisKey = thisKeyName, OtherKey = otherKeyName, CanBeNull = thisMember.IsNullable() });
    }

    public PropertyMappingBuilder<TEntity, IEnumerable<TPropElement>> Association<TPropElement, TThisKey, TOtherKey>(
      Expression<Func<TEntity, IEnumerable<TPropElement>>> prop,
      Expression<Func<TEntity, TThisKey>> thisKey,
      Expression<Func<TPropElement, TOtherKey>> otherKey
      //bool canBeNull = true
      ) {
      if (prop == null) throw new ArgumentNullException(nameof(prop));
      if (thisKey == null) throw new ArgumentNullException(nameof(thisKey));
      if (otherKey == null) throw new ArgumentNullException(nameof(otherKey));
      var thisMember = MemberHelper.GetMemberInfo(thisKey);
      var thisKeyName = thisMember.Name;
      var otherKeyName = MemberHelper.GetMemberInfo(otherKey).Name;
      return Property(prop).HasAttribute(new AssociationAttribute { ThisKey = thisKeyName, OtherKey = otherKeyName, CanBeNull = thisMember.IsNullable() });
    }

    public PropertyMappingBuilder<TEntity, IEnumerable<TOther>> Association<TOther>(
      Expression<Func<TEntity, IEnumerable<TOther>>> prop,
      Expression<Func<TEntity, TOther, bool>> predicate
      ///bool canBeNull = true
      ) {
      if (prop == null) throw new ArgumentNullException(nameof(prop));
      if (predicate == null) throw new ArgumentNullException(nameof(predicate));
      return Property(prop).HasAttribute(new AssociationAttribute { Predicate = predicate, CanBeNull = MemberHelper.GetMemberInfo(prop).IsNullable(), Relationship = Relationship.OneToMany });
    }

    public PropertyMappingBuilder<TEntity, TOther> Association<TOther>(
      Expression<Func<TEntity, TOther>> prop,
      Expression<Func<TEntity, TOther, bool>> predicate
      //bool canBeNull = true
      ) {
      if (prop == null) throw new ArgumentNullException(nameof(prop));
      if (predicate == null) throw new ArgumentNullException(nameof(predicate));
      return Property(prop).HasAttribute(new AssociationAttribute { Predicate = predicate, CanBeNull = MemberHelper.GetMemberInfo(prop).IsNullable() });
    }

    public PropertyMappingBuilder<TEntity, IEnumerable<TOther>> Association<TOther>(
      Expression<Func<TEntity, IEnumerable<TOther>>> prop,
      Expression<Func<TEntity, IDataContext, IQueryable<TOther>>> queryExpression
      //bool canBeNull = true
      ) {
      if (prop == null) throw new ArgumentNullException(nameof(prop));
      if (queryExpression == null) throw new ArgumentNullException(nameof(queryExpression));
      return Property(prop).HasAttribute(new AssociationAttribute { QueryExpression = queryExpression, CanBeNull = MemberHelper.GetMemberInfo(prop).IsNullable(), Relationship = Relationship.OneToMany });
    }

    public PropertyMappingBuilder<TEntity, TOther> Association<TOther>(
      Expression<Func<TEntity, TOther>> prop,
      Expression<Func<TEntity, IDataContext, IQueryable<TOther>>> queryExpression
      //bool canBeNull = true
      ) {
      if (prop == null) throw new ArgumentNullException(nameof(prop));
      if (queryExpression == null) throw new ArgumentNullException(nameof(queryExpression));
      return Property(prop).HasAttribute(new AssociationAttribute { QueryExpression = queryExpression, CanBeNull = MemberHelper.GetMemberInfo(prop).IsNullable() });
    }
```
This would both make for less coding & could improve the quality of the code by negating the possibility of mistakenly using the wrong value for the **canBeNull** argument.

**Comment by travbeamo** [comment]
Basic summary is if the helper methods & code refactoring proposed above are accepted in to Linq2DB, then the canBeNull attribute could possibly be removed altogether.

Then the correct joins would be used purely based on the if the association properties in the entities model are nullable:

If association property is nullable = LEFT join is used
If association property is not nullable = INNER join is used

**Comment by sdanyliv** [comment]
I think it will be regression. I have removed implicit join detection from associations, so `CanBeNull = false` will generate inner join.

**Comment by MaceWindu** [comment]
IMplemented #3606
